app/models/ems_refresh/save_inventory_middleware.rb

```diff
@@ -59,7 +57,8 @@ def save_middleware_servers_inventory(ems, hashes, target = nil)
               end
 
     hashes.each do |h|
-      h[:server_group_id] = h.fetch_path(:middleware_server_group, :id)
+      nativeid = h.fetch_path(:middleware_server_group, :nativeid)
+      h[:server_group_id] = MiddlewareServerGroup.find_by_nativeid(nativeid).try(:id) unless nativeid.nil?
```

Comment:
the `save_inventory_middleware.rb` expect all the data to be in one big hash and that parents contains the children in that hash. If that holds, the convention is to process the data top-down.

The "go down step" is done by calling the magic `save_inventory_multi`. This function internally calls the `save_inventory_with_findkey` that [keeps track](https://git.io/viL7K) of the newly created entities and their ids, so that things can be properly wired together.

This works fine if entity has always 1 parent, however, servers are special. So the code basically finds the newly created server group by `nativeid` (we have the nativeid in tha hash, and it's guaranteed to be unique) and then we get the `id` from the server group entity. The id of group is required for creating the association between the server and group.

TL;DR I found more readable finding the id myself than re-using the magic infrastructure that has very strong assumption on the data :]